elide-core/src/main/java/com/yahoo/elide/parsers/expression/FilterExpressionVisitor.java

```diff
@@ -0,0 +1,130 @@
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+
+package com.yahoo.elide.parsers.expression;
+
+import com.yahoo.elide.core.CheckInstantiator;
+import com.yahoo.elide.core.EntityDictionary;
+import com.yahoo.elide.core.exceptions.IllegalPermissionExpressionException;
+import com.yahoo.elide.core.filter.expression.AndFilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.OrFilterExpression;
+import com.yahoo.elide.core.filter.expression.Visitor;
+import com.yahoo.elide.generated.parsers.ExpressionBaseVisitor;
+import com.yahoo.elide.generated.parsers.ExpressionParser;
+import com.yahoo.elide.security.FilterExpressionCheck;
+import com.yahoo.elide.security.RequestScope;
+import com.yahoo.elide.security.checks.Check;
+import com.yahoo.elide.security.checks.UserCheck;
+
+
+/**
+ * FilterExpressionVisitor parses a permission parseTree and returns the corresponding FilterExpression representation
+ * of it. This allows passing a security permission predicate down to datastore level to reduce in-memory permission
+ * verification workload.
+ * A few cases is not allow and will throw exception:
+ *      1. User define FilterExpressionCheck which returns null in getFilterExpression function.
+ *      2. User put a FilterExpressionCheck with a non-userCheck type check in OR relation.
+ */
+public class FilterExpressionVisitor extends ExpressionBaseVisitor<FilterExpression> implements CheckInstantiator {
+    private final EntityDictionary dictionary;
+    private final RequestScope requestScope;
+
+    public static final FilterExpression NO_EVALUATION_EXPRESSION = new FilterExpression() {
+        @Override
+        public <T> T accept(Visitor<T> visitor) {
+            return null;
+        }
+
+        @Override
+        public boolean isDirty() {
+            return true;
+        }
+    };
+    public static final FilterExpression USER_CHECK_EXPRESSION = new FilterExpression() {
+        @Override
+        public <T> T accept(Visitor<T> visitor) {
+            return null;
+        }
+
+        @Override
+        public boolean isDirty() {
+            return true;
+        }
+    };
+
+    public FilterExpressionVisitor(EntityDictionary dictionary, RequestScope requestScope) {
+        this.dictionary = dictionary;
+        this.requestScope = requestScope;
+    }
+
+    @Override
+    public FilterExpression visitOR(ExpressionParser.ORContext ctx) {
+        FilterExpression left = visit(ctx.left);
+        FilterExpression right = visit(ctx.right);
+        if (left.isDirty() && right.isDirty()) {
+            //case where both side don't contain FilterExpressionCheck
+            return NO_EVALUATION_EXPRESSION;
+        }
+        if (left == NO_EVALUATION_EXPRESSION || right == NO_EVALUATION_EXPRESSION) {
+            throw new IllegalPermissionExpressionException(ctx.left.getText(), ctx.right.getText());
```

Comment:
Just return NO_EVALUATION_EXPRESSION here - and let the caller of the visitor throw an exception if needed.